File: Algorithms/QuickSort/QuickSort.py
def main():
    lines_count, poins_count = [int(input_line) for input_line in input().split()]

    lines = []
    for i in range(0, lines_count):
        left, right = [int(input_line) for input_line in input().split()]
        lines.append([left, -1])
        lines.append([right, 1])

    points = [int(input_line) for input_line in input().split()]
    for i in range(poins_count):
        lines.append([points[i], 0, i])

    lines.sort()
    result = [0] * poins_count
    current = 0

    for i in range(len(lines)):
        if lines[i][1] == 1:
            current += -1
        if lines[i][1] == -1:
            current -= -1
        if lines[i][1] == 0:
            result[lines[i][2]] = current

    print(*result)


if __name__ == '__main__':
    main()

File: Algorithms/QuickSort/QuickSort.cpp
#include <vector>
#include <set>
#include <algorithm>
#include <numeric>
#include <iostream>


using Line = std::pair<int, int>;


int main() {
    unsigned linesCount, poinsCount;
    std::cin >> linesCount >> poinsCount;

    std::vector<Line> lines(linesCount);
    for (Line &line : lines) {
        std::cin >> line.first >> line.second;
    }

    std::vector<int> points(poinsCount);
    std::vector<unsigned> counts(poinsCount);
    for (int &point : points) {
        std::cin >> point;
    }

    std::sort(lines.begin(), lines.end());

    std::vector<unsigned> pointsIndexes(poinsCount);
    std::iota(pointsIndexes.begin(), pointsIndexes.end(), 0u);
    std::sort(pointsIndexes.begin(), pointsIndexes.end(), [&](unsigned left, unsigned right) { return points[left] < points[right]; });

    auto sortedByEnds = [](const Line *lhs, const Line *rhs) { return lhs->second < rhs->second; };
    std::multiset<const Line *, decltype(sortedByEnds)> activeLines(sortedByEnds);

    auto lineIter = lines.begin();
    for (unsigned pointIndex : pointsIndexes) {
        int point = points[pointIndex];

        while (!activeLines.empty() && (*activeLines.begin())->second < point) {
            activeLines.erase(activeLines.begin());
        }

        for (; lineIter != lines.end() && lineIter->first <= point; ++lineIter) {
            if (lineIter->second >= point) {
                activeLines.insert(&*lineIter);
            }
        }

        counts[pointIndex] = activeLines.size();
    }

    for (unsigned count : counts) {
        std::cout << count << " ";
    }
    std::cout << std::endl;
}